Unable to configure JPA idempotent consumers

Depending on how things are configured you either get a warning or fatal exception caused by the following: 

```
Caused by: java.lang.NoClassDefFoundError: org/hibernate/proxy/HibernateProxy
        at java.lang.ClassLoader.defineClass1(Native Method)
        at java.lang.ClassLoader.defineClass(ClassLoader.java:760)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:497)
        at javassist.util.proxy.FactoryHelper.toClass2(FactoryHelper.java:182)
        at javassist.util.proxy.FactoryHelper.toClass(FactoryHelper.java:164)
        ... 39 more
Caused by: java.lang.ClassNotFoundException: org.hibernate.proxy.HibernateProxy from [Module "org.apache.camel.component.jpa:main" from local module loader @704d6e83 (finder: local module finder @43a0cee9]
```

--- a/feature-pack/src/main/resources/modules/system/layers/fuse/org/apache/camel/component/jpa/main/module.xml
+++ b/feature-pack/src/main/resources/modules/system/layers/fuse/org/apache/camel/component/jpa/main/module.xml
@@ -5,6 +5,8 @@
   </resources>
   <dependencies>
     <module name="javax.api" />
+    <module name="org.hibernate" />
+    <module name="org.javassist" />
     <module name="org.springframework.orm" export="true">
       <exports>
         <include path="org/springframework/orm/jpa" />

--- a/itests/standalone/basic/src/main/java/org/wildfly/camel/test/jpa/JPAIdempotentConsumerIntegrationTest.java
+++ b/itests/standalone/basic/src/main/java/org/wildfly/camel/test/jpa/JPAIdempotentConsumerIntegrationTest.java
@@ -1,0 +1,97 @@
+/*
+ * #%L
+ * Wildfly Camel :: Testsuite
+ * %%
+ * Copyright (C) 2013 - 2015 RedHat
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+
+package org.wildfly.camel.test.jpa;
+
+import javax.annotation.Resource;
+import javax.persistence.EntityManagerFactory;
+import javax.transaction.UserTransaction;
+
+import org.apache.camel.ProducerTemplate;
+import org.apache.camel.builder.RouteBuilder;
+import org.apache.camel.component.mock.MockEndpoint;
+import org.apache.camel.impl.DefaultCamelContext;
+import org.apache.camel.processor.idempotent.jpa.JpaMessageIdRepository;
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.arquillian.junit.Arquillian;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.asset.EmptyAsset;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.transaction.TransactionDefinition;
+import org.springframework.transaction.jta.JtaTransactionManager;
+import org.springframework.transaction.support.TransactionTemplate;
+import org.wildfly.extension.camel.CamelAware;
+
+@CamelAware
+@RunWith(Arquillian.class)
+public class JPAIdempotentConsumerIntegrationTest {
+
+    @Resource(mappedName = "java:jboss/IdempotentTestManagerFactory")
+    private EntityManagerFactory entityManagerFactory;
+
+    @Resource(mappedName = "java:jboss/UserTransaction")
+    private UserTransaction userTransaction;
+
+    @Deployment
+    public static JavaArchive deployment() {
+        JavaArchive archive = ShrinkWrap.create(JavaArchive.class, "jpa-idempotent-tests.jar");
+        archive.addAsResource("jpa/persistence-idempotent.xml", "META-INF/persistence.xml");
+        archive.addAsManifestResource(EmptyAsset.INSTANCE, "beans.xml");
+        return archive;
+    }
+
+    @Test
+    public void testJPAIdempotentConsumer() throws Exception {
+        final TransactionTemplate transactionTemplate = new TransactionTemplate();
+        transactionTemplate.setTransactionManager(new JtaTransactionManager(userTransaction));
+        transactionTemplate.setPropagationBehavior(TransactionDefinition.PROPAGATION_REQUIRED);
+
+        final JpaMessageIdRepository messageIdRepository = new JpaMessageIdRepository(entityManagerFactory, transactionTemplate, "myProcessorName");
+
+        DefaultCamelContext camelctx = new DefaultCamelContext();
+        camelctx.addRoutes(new RouteBuilder() {
+            @Override
+            public void configure() throws Exception {
+                from("direct:start")
+                .idempotentConsumer(simple("${header.messageId}"), messageIdRepository)
+                .to("mock:result");
+            }
+        });
+
+        camelctx.start();
+        try {
+            MockEndpoint mockEndpoint = camelctx.getEndpoint("mock:result", MockEndpoint.class);
+            mockEndpoint.expectedMessageCount(1);
+
+            // Send 5 messages with the same messageId header. Only 1 should be forwarded to the mock:result endpoint
+            ProducerTemplate template = camelctx.createProducerTemplate();
+            for (int i = 0; i < 5; i++) {
+                template.requestBodyAndHeader("direct:start", null, "messageId", "12345");
+            }
+
+            mockEndpoint.assertIsSatisfied();
+        } finally {
+            camelctx.stop();
+        }
+    }
+
+}

--- a/itests/standalone/basic/src/main/java/org/wildfly/camel/test/jpa/JPATransactionManagerIntegrationTest.java
+++ b/itests/standalone/basic/src/main/java/org/wildfly/camel/test/jpa/JPATransactionManagerIntegrationTest.java
@@ -37,7 +37,7 @@ import org.wildfly.camel.test.jpa.subA.Account;
 import org.wildfly.extension.camel.CamelContextRegistry;
 
 @RunWith(Arquillian.class)
-public class JpaTransactionManagerIntegrationTest {
+public class JPATransactionManagerIntegrationTest {
 
     @ArquillianResource
     CamelContextRegistry contextRegistry;

--- a/itests/standalone/basic/src/main/java/org/wildfly/camel/test/jpa/UserManagedTransactionIntegrationTest.java
+++ b/itests/standalone/basic/src/main/java/org/wildfly/camel/test/jpa/UserManagedTransactionIntegrationTest.java
@@ -47,7 +47,7 @@ import org.wildfly.camel.test.jpa.subA.JpaRouteBuilder;
 import org.wildfly.extension.camel.CamelContextRegistry;
 
 @RunWith(Arquillian.class)
-public class TransactionalJpaIntegrationTest {
+public class UserManagedTransactionIntegrationTest {
 
     @ArquillianResource
     CamelContextRegistry contextRegistry;

--- a/itests/standalone/basic/src/main/resources/jpa/persistence-idempotent.xml
+++ b/itests/standalone/basic/src/main/resources/jpa/persistence-idempotent.xml
@@ -1,0 +1,14 @@
+<persistence version="2.0"
+             xmlns="http://java.sun.com/xml/ns/persistence" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+             xsi:schemaLocation="
+        http://java.sun.com/xml/ns/persistence
+        http://java.sun.com/xml/ns/persistence/persistence_2_0.xsd">
+    <persistence-unit name="idempotentDb">
+        <jta-data-source>java:jboss/datasources/ExampleDS</jta-data-source>
+        <class>org.apache.camel.processor.idempotent.jpa.MessageProcessed</class>
+        <properties>
+            <property name="jboss.entity.manager.factory.jndi.name" value="java:jboss/IdempotentTestManagerFactory" />
+            <property name="hibernate.hbm2ddl.auto" value="create-drop"/>
+        </properties>
+    </persistence-unit>
+</persistence>

--- a/modules/etc/generated/wildfly/modules/system/layers/fuse/org/apache/camel/component/jpa/main/module.xml
+++ b/modules/etc/generated/wildfly/modules/system/layers/fuse/org/apache/camel/component/jpa/main/module.xml
@@ -5,6 +5,8 @@
   </resources>
   <dependencies>
     <module name="javax.api" />
+    <module name="org.hibernate" />
+    <module name="org.javassist" />
     <module name="org.springframework.orm" export="true">
       <exports>
         <include path="org/springframework/orm/jpa" />

--- a/modules/etc/smartics/camel-modules.xml
+++ b/modules/etc/smartics/camel-modules.xml
@@ -454,6 +454,8 @@
         </apply-to-dependencies>
         <dependencies>
             <module name="javax.api"/>
+            <module name="org.hibernate" />
+            <module name="org.javassist" />
             <module name="org.springframework.orm" export="true">
                 <exports>
                     <include path="org/springframework/orm/jpa" />

